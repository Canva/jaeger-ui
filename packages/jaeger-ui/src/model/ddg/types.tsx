// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TVertex } from '@jaegertracing/plexus/lib/types';

import PathElem from './PathElem';
import { fetchedState } from '../../constants';
import { ApiError } from '../../types/api-error';

export { default as PathElem } from './PathElem';

export type TDdgPayloadEntry = {
  operation: string;
  service: string;
};

export type TDdgPayload = TDdgPayloadEntry[][];

export type TDdgService = {
  name: string;
  operations: Map<string, TDdgOperation>;
};

export type TDdgOperation = {
  name: string;
  pathElems: PathElem[];
  service: TDdgService;
};

export type TDdgServiceMap = Map<string, TDdgService>;

export type TDdgPath = {
  focalIdx: number;
  members: PathElem[];
};

export type TDdgDistanceToPathElems = Map<number, PathElem[]>;

export type TDdgModel = {
  distanceToPathElems: TDdgDistanceToPathElems;
  paths: TDdgPath[];
  services: TDdgServiceMap;
  visIdxToPathElem: PathElem[];
};

export type TDdgVertex = TVertex<{ service: string; operation: string }>;

export type TDdgStateEntry =
  | {
      state: typeof fetchedState.LOADING;
    }
  | {
      error: ApiError;
      state: typeof fetchedState.ERROR;
    }
  | {
      model: TDdgModel;
      state: typeof fetchedState.DONE;
      styleStates: Map<number, number>;
    };

export type TDdgState = Record<string, Record<string, Record<string, Record<string, TDdgStateEntry>>>>;

export enum EViewModifier {
  None,
  Hovered,
  Selected,
  Emphasized = 1 << 2, // eslint-disable-line no-bitwise
}

export type TDdgRequiredUrlState = {
  service: string;
  operation?: string;
  start: number;
  end: number;
};

export type TDdgActionMeta = {
  query: TDdgRequiredUrlState;
};

export type TDdgAddStylePayload = TDdgRequiredUrlState & { visibilityIndices: number[]; style: number };

export type TDdgClearStylePayload = TDdgRequiredUrlState & { visibilityIndices?: number[]; style?: number };
