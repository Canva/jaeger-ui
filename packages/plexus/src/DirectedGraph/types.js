// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import type { Edge, LayoutEdge, LayoutGraph, LayoutVertex, SizeVertex, Vertex } from '../types/layout';

import LayoutManager from '../LayoutManager';

export type DirectedGraphProps = {
  classNamePrefix: string,
  edges: Edge[],
  getEdgeLabel: Edge => string | React.Node,
  getNodeLabel: Vertex => string | React.Node,
  layoutManager: LayoutManager,
  // setOnEdgeArrow
  setOnEdgePath: ?(Edge) => {},
  setOnEdgesContainer: ?(?LayoutGraph) => {},
  setOnNode: ?(Vertex) => {},
  setOnNodesContainer: ?(?LayoutGraph) => {},
  setOnRoot: ?(?LayoutGraph) => {},
  vertices: Vertex[],
};

export type DirectedGraphState = {
  edges: Edge[],
  layoutEdges: ?(LayoutEdge[]),
  layoutGraph: ?LayoutGraph,
  layoutPhase: number,
  layoutVertices: ?(LayoutVertex[]),
  sizeVertices: ?(SizeVertex[]),
  vertexRefs: { current: ?HTMLElement }[],
  vertices: Vertex[],
};
