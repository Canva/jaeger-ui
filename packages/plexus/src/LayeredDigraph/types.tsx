// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { TEdge, TLayoutEdge, TLayoutGraph, TLayoutVertex, TVertex } from '../types';
import { TOneOfFour, TOneOfTwo } from '../types/TOneOf';
import { ZoomTransform } from '../ZoomManager';

export enum ELayoutPhase {
  NoData = 'NoData',
  CalcSizes = 'CalcSizes',
  CalcPositions = 'CalcPositions',
  CalcEdges = 'CalcEdges',
  Done = 'Done',
}

export type TLayerType = 'html' | 'svg';
export enum ELayerType {
  Html = 'html',
  Svg = 'svg',
}

export type TRendererUtils = {
  getLocalId: (name: string) => string;
  getZoomTransform: () => ZoomTransform;
};

export type TExposedGraphState<T = {}, U = {}> = {
  edges: TEdge<U>[];
  layoutEdges: TLayoutEdge<U>[] | null;
  layoutGraph: TLayoutGraph | null;
  layoutPhase: ELayoutPhase;
  layoutVertices: TLayoutVertex<T>[] | null;
  renderUtils: TRendererUtils;
  vertices: TVertex<T>[];
  zoomTransform: ZoomTransform;
};

export type TAnyProps = Record<string, unknown> & {
  className?: string;
  style?: React.CSSProperties;
};

export type TPropFactoryFn = (...args: any[]) => TAnyProps | null;

export type TSetProps<TFactoryFn extends TPropFactoryFn> =
  | TAnyProps
  | TFactoryFn
  | (TAnyProps | TFactoryFn)[];

export type TFromGraphStateFn<T = {}, U = {}> = (input: TExposedGraphState<T, U>) => TAnyProps | null;

export type TPropsFactory<PropNames extends string, FactoryFn extends TPropFactoryFn> = {
  [K in PropNames]?: TSetProps<FactoryFn>;
};

export type TSetOnContainer<T = {}, U = {}> = TPropsFactory<'setOnContainer', TFromGraphStateFn<T, U>>;

type TKeyed = { key: string };

export type TNodeRenderFn<T = {}> = (vertex: TVertex<T>, utils: TRendererUtils) => React.ReactNode;

export type TMeasurableNodeRenderFn<T = {}> = (
  vertex: TVertex<T>,
  utils: TRendererUtils,
  layoutVertex: TLayoutVertex<T> | null
) => React.ReactNode;

export type TMeasurableNodeRenderer<T = {}> = {
  measurable: true;
  nodeRender: TMeasurableNodeRenderFn<T>;
  setOnNode?: TSetProps<
    (vertex: TVertex<T>, utils: TRendererUtils, layoutVertex: TLayoutVertex<T> | null) => TAnyProps | null
  >;
};

type TNodeRenderer<T = {}> = {
  nodeRender: TNodeRenderFn<T>;
  setOnNode?: TSetProps<(layoutVertex: TLayoutVertex<T>, utils: TRendererUtils) => TAnyProps | null>;
};

type TNodesLayer<T = {}, U = {}> = TKeyed &
  TOneOfTwo<TNodeRenderer<T>, TMeasurableNodeRenderer<T>> &
  TSetOnContainer<T, U>;

type TStandaloneNodesLayer<T = {}, U = {}> = TNodesLayer<T, U> & {
  layerType: TLayerType;
};

type TMarkerDef<T = {}, U = {}> = TKeyed & {
  type: React.Component;
  localId: string;
  setOnMarker?: TSetProps<TFromGraphStateFn<T, U>>;
};

type TEdgesLayer<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    edges: true;
    setOnEdge?: TSetProps<(edges: TLayoutEdge<U>, utils: TRendererUtils) => TAnyProps | null>;
  };

type TStandaloneEdgesLayer<T = {}, U = {}> = TEdgesLayer<T, U> & {
  layerType: TLayerType;
  defs?: TMarkerDef<T, U>[];
};

export type THtmlLayersGroup<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    layerType: Extract<TLayerType, 'html'>;
    layers: TOneOfTwo<TNodesLayer<T, U>, TEdgesLayer<T, U>>[];
  };

type TSvgLayersGroup<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    layerType: Extract<TLayerType, 'svg'>;
    defs?: TMarkerDef<T, U>[];
    layers: TOneOfTwo<TNodesLayer<T, U>, TEdgesLayer<T, U>>[];
  };

export type TLayer<T = {}, U = {}> = TOneOfFour<
  THtmlLayersGroup<T, U>,
  TSvgLayersGroup<T, U>,
  TStandaloneNodesLayer<T, U>,
  TStandaloneEdgesLayer<T, U>
>;

export type TMeasurableNodeProps<T = {}> = Omit<TMeasurableNodeRenderer<T>, 'measurable'> & {
  classNamePrefix?: string;
  hidden?: boolean;
  layoutVertex: TLayoutVertex<T> | null;
  renderUtils: TRendererUtils;
  vertex: TVertex<T>;
};
