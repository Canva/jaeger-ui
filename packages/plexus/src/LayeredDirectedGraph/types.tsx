// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { TEdge, TLayoutEdge, TLayoutGraph, TLayoutVertex, TVertex } from '../types';
import TOneOf from '../types/TOneOf';
import { ZoomTransform } from '../ZoomManager';

export enum ELayoutPhase {
  NoData = 'NoData',
  CalcSizes = 'CalcSizes',
  CalcPositions = 'CalcPositions',
  CalcEdges = 'CalcEdges',
  Done = 'Done',
}

export type TAnyProps = Record<string, unknown> & {
  className?: string;
  style?: React.CSSProperties;
};

export type TPropFactoryFn = (...args: any[]) => TAnyProps | null;

export type TSetProps<TFactoryFn extends TPropFactoryFn> =
  | TAnyProps
  | TFactoryFn
  | (TAnyProps | TFactoryFn)[];

// TODO(joe): consider getting rid of this type
// type TAesthetics = {
//   className?: string;
//   style?: React.CSSProperties;
// };

export type TRendererUtils = {
  getLocalId: (name: string) => string;
  getZoomTransform: () => ZoomTransform;
};

export type TExposedGraphState<T = {}, U = {}> = {
  edges: TEdge<U>[];
  zoomTransform: ZoomTransform;
  layoutEdges: TLayoutEdge<U>[] | null;
  layoutGraph: TLayoutGraph | null;
  layoutPhase: ELayoutPhase;
  layoutVertices: TLayoutVertex<T>[] | null;
  renderUtils: TRendererUtils;
  vertices: TVertex<T>[];
};

export type TFromGraphStateFn<T = {}, U = {}> = (input: TExposedGraphState<T, U>) => TAnyProps | null;

export type TSetOnContainer<T = {}, U = {}> = {
  setOnContainer?: TSetProps<TFromGraphStateFn<T, U>>;
};

type TKeyed = { key: string };

type TElemType = TOneOf<{ html: true }, { svg: true }>;

export type TNodeRenderFn<T = {}> = (vertex: TVertex<T>, utils: TRendererUtils) => React.ReactNode;

export type TMeasurableNodeRenderFn<T = {}> = (
  vertex: TVertex<T>,
  utils: TRendererUtils,
  layoutVertex: TLayoutVertex<T> | null
) => React.ReactNode;

type TMeasurableNodeRenderer<T = {}> = {
  measurable: true;
  nodeRender: TMeasurableNodeRenderFn<T>;
  setOnNode?: TSetProps<
    (vertex: TVertex<T>, utils: TRendererUtils, layoutVertex: TLayoutVertex<T> | null) => TAnyProps | null
  >;
};

type TNodeRenderer<T = {}> = {
  nodeRender: TNodeRenderFn<T>;
  setOnNode?: TSetProps<(layoutVertex: TLayoutVertex<T>, utils: TRendererUtils) => TAnyProps | null>;
};

type TNodesLayer<T = {}, U = {}> = TKeyed &
  TOneOf<TNodeRenderer<T>, TMeasurableNodeRenderer<T>> &
  TSetOnContainer<T, U>;

type TStandaloneNodesLayer<T = {}, U = {}> = TNodesLayer<T, U> & TElemType;

type TEdgesLayer<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    edges: true;
    setOnEdge?: TSetProps<(edges: TLayoutEdge<U>, utils: TRendererUtils) => TAnyProps | null>;
  };

type TStandaloneEdgesLayer<T = {}, U = {}> = TEdgesLayer<T, U> & TElemType;

type THtmlLayersGroup<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    html: true;
    layers: (TNodesLayer<T, U> | TEdgesLayer<T, U>)[];
  };

type TMarkerDef<T = {}, U = {}> = TKeyed & {
  type: React.Component;
  localId: string;
  setOnMarker?: TSetProps<TFromGraphStateFn<T, U>>;
};

type TSvgLayersGroup<T = {}, U = {}> = TKeyed &
  TSetOnContainer<T, U> & {
    svg: true;
    defs?: TMarkerDef<T, U>[];
    layers: (TNodesLayer<T, U> | TEdgesLayer<T, U>)[];
  };

export type TLayer<T = {}, U = {}> = TOneOf<
  THtmlLayersGroup<T, U>,
  TSvgLayersGroup<T, U>,
  TStandaloneNodesLayer<T, U>,
  TStandaloneEdgesLayer<T, U>
>;
