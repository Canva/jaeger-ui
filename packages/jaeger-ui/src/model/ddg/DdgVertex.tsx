// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TVertex } from '@jaegertracing/plexus/lib/types';

import { DdgEdge, PathElem } from './types';

type TDdgVertex = {
  egressEdges: Map<DdgVertex, DdgEdge>;
  ingressEdges: Map<DdgVertex, DdgEdge>;
  pathElems: Set<PathElem>;
};

export default class DdgVertex implements TVertex<TDdgVertex> {
  egressEdges: Map<DdgVertex, DdgEdge>;
  key: string;
  ingressEdges: Map<DdgVertex, DdgEdge>;
  pathElems: Set<PathElem>;

  constructor({ key }: { key: string }) {
    this.egressEdges = new Map();
    this.key = key;
    this.ingressEdges = new Map();
    this.pathElems = new Set();
  }

  toJSON() {
    type TDigestibleEdges = Record<string, DdgEdge>;
    const digestibleEgressEdges: TDigestibleEdges = {};
    const digestibleIngressEdges: TDigestibleEdges = {};
    this.egressEdges.forEach((edge, egreesNeighbor) => {
      digestibleEgressEdges[egreesNeighbor.key] = edge;
    });
    this.ingressEdges.forEach((edge, ingreesNeighbor) => {
      digestibleIngressEdges[ingreesNeighbor.key] = edge;
    });
    return {
      egressEdges: digestibleEgressEdges,
      key: this.key,
      ingressEdges: digestibleIngressEdges,
      pathElems: this.pathElems,
    };
  }

  toString() {
    return JSON.stringify(this.toJSON(), null, 2);
  }

  get [Symbol.toStringTag]() {
    return `DdgVertex ${this.key}`;
  }
}
