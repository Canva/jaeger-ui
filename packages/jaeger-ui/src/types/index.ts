// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// TODO: Everett is this correct? Can't find ContextRouter
import { Router } from 'react-router-dom'; // eslint-disable-line no-unused-vars
import { History as RouterHistory, Location } from 'history'; // eslint-disable-line no-unused-vars

import { ApiError } from './api-error'; // eslint-disable-line no-unused-vars
import { TracesArchive } from './archive'; // eslint-disable-line no-unused-vars
import { Config } from './config'; // eslint-disable-line no-unused-vars
import { EmbeddedState } from './embedded'; // eslint-disable-line no-unused-vars
import TNullable from './nullable'; // eslint-disable-line no-unused-vars
import { SearchQuery } from './search'; // eslint-disable-line no-unused-vars
import { Trace } from './trace'; // eslint-disable-line no-unused-vars
import { TraceDiffState } from './trace-diff'; // eslint-disable-line no-unused-vars
import { TraceTimeline } from './trace-timeline'; // eslint-disable-line no-unused-vars

export type FetchedState = 'FETCH_DONE' | 'FETCH_ERROR' | 'FETCH_LOADING';

export type FetchedTrace = {
  data?: Trace;
  error?: ApiError;
  id: string;
  state?: FetchedState;
};

export type ReduxState = {
  archive: TracesArchive;
  config: Config;
  dependencies: {
    dependencies: { parent: string; child: string; callCount: number }[];
    loading: boolean;
    error: ApiError | TNullable;
  };
  embedded: EmbeddedState;
  router: Router & {
    location: Location;
  };
  services: {
    services: (string[]) | TNullable;
    operationsForService: Record<string, string[]>;
    loading: boolean;
    error: ApiError | TNullable;
  };
  trace: {
    traces: Record<string, FetchedTrace>;
    search: {
      error?: ApiError;
      results: string[];
      state?: FetchedState;
      query?: SearchQuery;
    };
  };
  traceDiff: TraceDiffState;
  traceTimeline: TraceTimeline;
};
