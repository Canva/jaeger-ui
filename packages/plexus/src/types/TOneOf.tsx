// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export type TOneOf<A, B, C = {}, D = {}, E = {}> =
  | { [P in Exclude<keyof (B & C & D & E), keyof A>]?: null } & A
  | { [P in Exclude<keyof (A & C & D & E), keyof B>]?: null } & B
  | { [P in Exclude<keyof (A & B & D & E), keyof C>]?: null } & C
  | { [P in Exclude<keyof (A & B & C & E), keyof D>]?: null } & D
  | { [P in Exclude<keyof (A & B & C & D), keyof E>]?: null } & E;

// eslint-disable-next-line no-undef
export default TOneOf;

export type TOneOfTwo<A, B> =
  | { [P in Exclude<keyof B, keyof A>]?: null } & A
  | { [P in Exclude<keyof A, keyof B>]?: null } & B;

export type TOneOfThree<A, B, C> =
  | { [P in Exclude<keyof (B & C), keyof A>]?: null } & A
  | { [P in Exclude<keyof (A & C), keyof B>]?: null } & B
  | { [P in Exclude<keyof (A & B), keyof C>]?: null } & C;

export type TOneOfFour<A, B, C, D> =
  | { [P in Exclude<keyof (B & C & D), keyof A>]?: null } & A
  | { [P in Exclude<keyof (A & C & D), keyof B>]?: null } & B
  | { [P in Exclude<keyof (A & B & D), keyof C>]?: null } & C
  | { [P in Exclude<keyof (A & B & C), keyof D>]?: null } & D;

export type TOneOfFive<A, B, C, D, E> = TOneOf<A, B, C, D, E>;
