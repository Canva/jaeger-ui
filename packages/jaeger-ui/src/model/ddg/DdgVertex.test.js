// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import DdgVertex from './DdgVertex';
import { DdgEdge } from './types';

describe('DdgVertex', () => {
  function makeTestVertex(vertexKey) {
    const vertex = new DdgVertex({ key: vertexKey });
    vertex.pathElems.add(`${vertexKey}-pathElem0`);
    vertex.pathElems.add(`${vertexKey}-pathElem1`);
    return vertex;
  }

  function makeTestEdge(from, to) {
    const edge = new DdgEdge({ from, to });
    from.egressEdges.set(to, edge);
    to.ingressEdges.set(from, edge);
  }

  const twoHopUpstreamVertexKey = 'twoHopUpstreamVertexKey';
  const oneHopUpstreamVertexKey0 = 'oneHopUpstreamVertexKey0';
  const oneHopUpstreamVertexKey1 = 'oneHopUpstreamVertexKey1';
  const targetVertexKey = 'targetVertexKey';
  const siblingVertexKey = 'siblingVertexKey';
  const oneHopDownstreamVertexKey0 = 'oneHopDownstreamVertexKey0';
  const oneHopDownstreamVertexKey1 = 'oneHopDownstreamVertexKey1';
  const twoHopDownstreamVertexKey = 'twoHopDownstreamVertexKey';

  const twoHopUpstreamVertex = makeTestVertex(twoHopUpstreamVertexKey);
  const oneHopUpstreamVertex0 = makeTestVertex(oneHopUpstreamVertexKey0);
  const oneHopUpstreamVertex1 = makeTestVertex(oneHopUpstreamVertexKey1);
  const targetVertex = makeTestVertex(targetVertexKey);
  const siblingVertex = makeTestVertex(siblingVertexKey);
  const oneHopDownstreamVertex0 = makeTestVertex(oneHopDownstreamVertexKey0);
  const oneHopDownstreamVertex1 = makeTestVertex(oneHopDownstreamVertexKey1);
  const twoHopDownstreamVertex = makeTestVertex(twoHopDownstreamVertexKey);

  const twoHopToOneHop0UpstreamEdge = makeTestEdge(twoHopUpstreamVertex, oneHopUpstreamVertex0);
  const oneHop0ToTargetUpstreamEdge = makeTestEdge(oneHopUpstreamVertex0, targetVertex);
  const targetToOneHop0DownstreamEdge = makeTestEdge(targetVertex, oneHopDownstreamVertex0);
  const oneHop0ToTwoHopDownstreamEdge = makeTestEdge(oneHopDownstreamVertex0, twoHopDownstreamVertex);

  const oneHop1ToTargetUpstreamEdge = makeTestEdge(oneHopUpstreamVertex1, targetVertex);
  const targetToOneHop1DownstreamEdge = makeTestEdge(targetVertex, oneHopDownstreamVertex1);

  const oneHop1ToSiblingUpstreamEdge = makeTestEdge(oneHopUpstreamVertex1, siblingVertex);
  const siblingToOneHop1DownstreamEdge = makeTestEdge(siblingVertex, oneHopDownstreamVertex1);

  it('initializes', () => {
    const vertexKey = 'vertex-key';
    const testDdgVertex = new DdgVertex({ key: vertexKey });
    expect(testDdgVertex.egressEdges).toEqual(new Map());
    expect(testDdgVertex.key).toEqual(vertexKey);
    expect(testDdgVertex.ingressEdges).toEqual(new Map());
    expect(testDdgVertex.pathElems).toEqual(new Set());
  });

  /*
  it('creates partial JSON', () => {
    const partialJSON = targetVertex.toJSONHelper();
    expect(partialJSON).toEqual({
      key: targetVertexKey,
      pathElems: targetVertex.pathElems,
    });
  });
  */

  it('creates consumable JSON', () => {
    expect(targetVertex.toJSON()).toMatchSnapshot();
  });

  it('creates consumable string', () => {
    expect(targetVertex.toString()).toBe(JSON.stringify(targetVertex.toJSON(), null, 2));
  });

  it('creates informative string tag', () => {
    expect(Object.prototype.toString.call(targetVertex)).toEqual(`[object DdgVertex ${targetVertexKey}]`);
  });
});
