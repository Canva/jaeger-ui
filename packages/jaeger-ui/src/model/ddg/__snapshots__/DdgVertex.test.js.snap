// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DdgVertex creates consumable JSON 1`] = `
Object {
  "egressVertices": Array [
    "oneHopDownstreamVertexKey0",
    "oneHopDownstreamVertexKey1",
  ],
  "ingressVertices": Array [
    "oneHopDownstreamVertexKey0",
    "oneHopDownstreamVertexKey1",
  ],
  "key": "targetVertexKey",
  "pathElems": Set {
    "targetVertexKey-pathElem0",
    "targetVertexKey-pathElem1",
  },
}
`;
